## Clustering

Last class we studied document vectors and how to find key words and similar documents. What else can we do with vectors? We can cluster them to find natural groups or categories, or visualize them directly by projecting them to 2D or 3D space.

In [1]:
import pandas as pd
import numpy as np
from textblob import TextBlob
from sklearn.feature_extraction.text import TfidfVectorizer

We'll start by getting tf-idf vectors for the Menendez press releases, like we did last class.

In [3]:
pr = pd.read_csv('week-2/menendez-press-releases.csv')
len(pr)

1530

In [4]:
# need a tokenizer
def tokenize(s):
  blob = TextBlob(s.lower())
  words = [token for token in blob.words if len(token) > 2]
  return words

In [5]:
# create tf-idf vectors
vectorizer = TfidfVectorizer(stop_words='english', tokenizer=tokenize)

matrix = vectorizer.fit_transform(pr.text)

# The easiest way to see what happenned is to make a dataframe
tfidf = pd.DataFrame(matrix.toarray(), columns=vectorizer.get_feature_names())
tfidf.head()

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>'01</th>
      <th>'activist</th>
      <th>'acts</th>
      <th>'disappointing</th>
      <th>'e-verify</th>
      <th>'em</th>
      <th>'liberty</th>
      <th>'ll</th>
      <th>'no</th>
      <th>'re</th>
      <th>...</th>
      <th>…oil-drilling</th>
      <th>…struggling</th>
      <th>…tarp</th>
      <th>…that</th>
      <th>…the</th>
      <th>…then</th>
      <th>…there</th>
      <th>…these</th>
      <th>…this</th>
      <th>…we</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>3</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>4</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 27656 columns</p>
</div>

We're going to use a clustering algorithm called k-means. Here's an interactive demo of how it works.
See this [interactive demo](http://web.stanford.edu/class/ee103/visualizations/kmeans/kmeans.html) or [this one](https://www.naftaliharris.com/blog/visualizing-k-means-clustering/).

In [6]:
from sklearn.cluster import KMeans

In [12]:
# cluster
number_of_clusters = 8
km = KMeans(n_clusters=number_of_clusters)
km.fit(matrix)

KMeans(algorithm='auto', copy_x=True, init='k-means++', max_iter=300,
    n_clusters=8, n_init=10, n_jobs=1, precompute_distances='auto',
    random_state=None, tol=0.0001, verbose=0)

Ok, let's see what's in each cluster!

In [9]:
km.cluster_centers_

array([[4.41800048e-04, 0.00000000e+00, 0.00000000e+00, ...,
        0.00000000e+00, 0.00000000e+00, 0.00000000e+00],
       [0.00000000e+00, 1.31855770e-04, 1.16727776e-04, ...,
        0.00000000e+00, 4.84559536e-05, 4.26850634e-05],
       [0.00000000e+00, 0.00000000e+00, 0.00000000e+00, ...,
        0.00000000e+00, 0.00000000e+00, 0.00000000e+00],
       [0.00000000e+00, 0.00000000e+00, 0.00000000e+00, ...,
        4.04502052e-04, 0.00000000e+00, 0.00000000e+00],
       [0.00000000e+00, 0.00000000e+00, 0.00000000e+00, ...,
        0.00000000e+00, 0.00000000e+00, 0.00000000e+00]])

In [14]:
km.labels_[:100]

array([4, 4, 7, 4, 5, 1, 1, 2, 4, 2, 2, 1, 2, 4, 2, 1, 4, 5, 2, 4, 3, 5,
       1, 4, 0, 4, 2, 1, 1, 1, 4, 4, 4, 1, 5, 1, 5, 4, 4, 7, 1, 2, 1, 1,
       0, 2, 1, 2, 1, 1, 0, 2, 2, 1, 5, 2, 1, 4, 1, 5, 2, 2, 7, 2, 1, 5,
       2, 4, 5, 2, 3, 2, 1, 5, 2, 1, 1, 5, 1, 4, 4, 2, 4, 1, 4, 4, 4, 4,
       4, 4, 1, 6, 6, 1, 5, 4, 4, 1, 1, 6], dtype=int32)

In [10]:
def print_sorted_vector(v):
    # this "lambda" thing is an anonymous function, google me to unluck bonus coding knowledge
    sorted_list = sorted(v.items(), key=lambda x: (x[1],x[0]), reverse=True) 
    sorted_list = sorted_list[:10]
    print('\n'.join([str(x) for x in sorted_list]))

Now we're going to print out the top words of the center vector of each cluster, to see how the k-means algorithm did.

In [13]:
# print cluster centroids
centroids = pd.DataFrame(km.cluster_centers_, columns=vectorizer.get_feature_names())
for i in range(number_of_clusters):
  print('Cluster', i)
  print_sorted_vector(centroids.iloc[i])

Cluster 0
('energy', 0.28408630967896437)
('efficiency', 0.09541028118649576)
('solar', 0.07742528705779031)
('program', 0.05738309594316209)
('clean', 0.05380111056996603)
('climate', 0.046672465509743195)
('funding', 0.04465339468527712)
('new', 0.04454287766652216)
('jersey', 0.04176444480098218)
('menendez', 0.04154694255381151)
Cluster 1


('new', 0.07403880239808748)
('jersey', 0.07239448245622229)
('funding', 0.06302642424111905)
('lautenberg', 0.055473228247399774)
('program', 0.04192095564587945)
('federal', 0.03941738932762954)
('transit', 0.03888714362979307)
('million', 0.03855502529699156)
('help', 0.037921770690117526)
('funds', 0.037708980717719456)
Cluster 2
('tax', 0.06182224836755691)
('credit', 0.054341825783588385)
('financial', 0.04813339667435814)
('street', 0.04797926810938029)
('wall', 0.0467416752179985)
('card', 0.038212918667974355)
('homeowners', 0.03725997218347599)
('menendez', 0.03662344958785023)
('small', 0.036497619473064395)
('businesses', 0.03433937513446772)
Cluster 3


('faa', 0.14720299722540706)
('airport', 0.12478163493219718)
('airports', 0.08241409117301865)
('fees', 0.08021447767918477)
('flight', 0.07204731782590147)
('airlines', 0.06971933752190458)
('airline', 0.06896441184452247)
('aviation', 0.06324404898522548)
('airfares', 0.06290100593930754)
('delays', 0.06005486616948002)
Cluster 4
('menendez', 0.034420073301741266)
('senator', 0.025316397610260062)
('new', 0.023238203240787927)
('foreign', 0.02202143289985424)
('iraq', 0.02098869678656269)
('u.s', 0.020909231014347305)
('statement', 0.01923583072950883)
('senate', 0.018454638584351705)
('women', 0.017865445326239238)
('security', 0.01783358791577416)
Cluster 5


('al-megrahi', 0.11460694853429186)
('release', 0.0943345908890778)
('regime', 0.08158882048205439)
('iran', 0.08036331305712162)
('scottish', 0.07432955356662534)
('libya', 0.06378528663557612)
('cuba', 0.06100700610956416)
('cuban', 0.059625460892580855)
('government', 0.053615165863569174)
('castro', 0.05360872134585864)
Cluster 6
('oil', 0.34494965340114175)
('drilling', 0.11844190446469596)
('companies', 0.10737204958680821)
('gas', 0.08958884662527034)
('spill', 0.07609009012658786)
('big', 0.06734134031481388)
('prices', 0.06632594503415469)
('energy', 0.05130964162295502)
('coast', 0.04614937203640204)
('billion', 0.044261202461053493)
Cluster 7


('health', 0.22095364859113995)
('care', 0.12803975688160713)
('children', 0.10286095999670188)
('insurance', 0.10111388512180315)
('autism', 0.07813159685996236)
('coverage', 0.06327098192301313)
('new', 0.06057796755439233)
('jersey', 0.04881731427158306)
('services', 0.046898764417348915)
('families', 0.03780864103885259)


In fact, Overview uses k-means in its "topic tree" visualization

### Visualizing clusters to understand politics
This is a fairly literal translation of a [previous post](http://www.compjournalism.com/?p=13) of mine (it was done in  R at the time). We're going to load up the voting record of the U.K. House of Lords, turn each MP's voting record into a vector, and see how all these politicians relate in this abstract ideological space.

The daia is circa 2012, because they had an interesting coalition government at the time. 

In [15]:
from sklearn.decomposition import PCA
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D
%matplotlib inline

In [ ]:
votes = pd.read_csv('week-2/uk-lords-votes.csv')
votes.shape

(613, 102)

In [9]:
# print out this votes matrix

This is data I processed earlier, and you can think of it as a template for the format you will need to get your data into to do your homework. Each row is one member of parliament. Each of the numbered columns is one vote, where 1 means aye, 0 means abstain, and -1 means nay. The `party` column indicates which political party that MP belonged to at the time.

If you're interested in the original data, including the names of these politicians and what they were voting on, you can find it all [here](http://www.compjournalism.com/?p=13).

We'll want to turn the list of parties in to a list of colors.

In [14]:
# compute the color that each MP should be, based on their party


Now that we've set everything up, we're ready to start projecting. We can view at most three dimensions at once with our puny human visual system. The simplest projection is just to pick three dimensions of our vectors and plot them.

In [8]:
# 3d scatterplot of three votes



Not very interesting. All of vote coordinates are in [-1,0,1] so no matter which votes (dimensions) we pick we can only get the corners, edges, and center of a cube. Plus, all 613 MPs overlap each other -- many MPs voted the same way on this set of three votes -- so we only see a few dots.

Instead, we're going to let the computer pick the right projection from this wacky high dimensional space to two dimensions. We are using PCA, "principal components analysis," which tries to find a direction to project that gives maximum separation of all the points. This dimension doesn't have to be aligned to any of our dimension axes -- PCA will "rotate" the points in high dimensional space until they are as spread out as possible.

In [5]:
# PCA to 2D

In [6]:
# 2D scatterplot

We can actually project down to any number of dimensions. More than 3 but less than the original 100 can be useful for some data processing operations.) Here, we'll project down to 3 and take a look at our voting clusters in glorious 3D.

In [18]:
# PCA to 3D

In [7]:
# 3D scatterplot